export suggestSimilar (incl. type definition)

jpilgrim opened the pull request:
# Pull Request

<!--
The text in these markdown comments is instructions that will not appear in the displayed pull request,
and can be deleted.

Please submit pull requests against the develop branch.

Follow the existing code style. Check the tests succeed, including lint.
  npm run test
  npm run lint

Don't update the CHANGELOG or command version number. That gets done by maintainers when preparing the release.

Commander currently has zero production dependencies. That isn't a hard requirement, but is a simple story. Requests which 
add a dependency are much less likely to be accepted, and we are likely to ask if there alternative approaches to avoid the dependency.
-->

## Problem

<!--
What problem are you solving?
What Issues does this relate to?
Show the broken output if appropriate.
-->

Similar to internal commander functionality I needed a suggestSimilar function in my client (using commander as a dependency).

## Solution

<!--
How did you solve the problem? 
Show the fixed output if appropriate.

There are a lot of forms of documentation which could need updating for a change in functionality. It
is ok if you want to show us the code to discuss before doing the extra work, and
you should say so in your comments so we focus on the concept first before talking about all the other pieces:

- TypeScript typings
- JSDoc documentation in code
- tests
- README
- examples/
-->

Instead of writing the function twice or using another dependency, I simply exported the suggestSimilar function already included in commander.js. That includes its type definitions.

This is a perfect solution, since I needed exactly the format produced by suggestSimilar -- my client is a CLI after all, which is why I use commander.js in the first place.

Just one note: I also updated readme.md and copied the section into readme_zh-CN.md (in english) to simplify translation.

## ChangeLog

<!--
Optional. Suggest a line for adding to the CHANGELOG to summarise your change.
-->

- export suggestSimilar to be used in client code as well

shadowspawn commented:
I am pleased that you think it is useful, but my initial reaction is that I do not want to expose the routine as it is tailored for internal use and not core to what Commander offers. There are multiple libraries that are focused on suggestions.

I'll leave it for a while and think about it some more.

(For interest, the sort of opposite direction I had considered was whether people would want to supply their own implementation. I decided it was not particularly likely so left out, and reconsider if requested and upvoted.)

jpilgrim commented:
@shadowspawn I understand your concerns. In order to make the function a little bit more generic (i.e., less "tailored for internal use") I have extracted the core part into a new function computeSimilar. I also added tests. Now, only the more general function computeSimilar is exposed.

Commander does not use any dependencies, which is great. For small command line tools, I also want to keep the number of dependencies small. Suggesting a similar word (e.g., for wrongly written filenames) is a typical CLI feature (in a GUI, this usually does not happen as the file is picked from a list). This is a very typical case for CLIs, and I would hate to add another dependency (or otherwise duplicate the solution already found in commander.

If someone needs a more specific solution, one can always use another packages as well.

shadowspawn commented:
Thanks for the suggestion. This is not something I wish to make public at this time.

I think this is the first time the suggestion code has come up for discussion. If there are future issues and demand this can be revisited.
